### PR TITLE
KAFKA-4986: Add producer per task support

### PR DESCRIPTION
**Pull request by mjsax**
Enable producer per task if exactly-once config is enabled.

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @enothereska @dguy @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:97]: Why so many changes in this class? Aren't we just adding a single config option?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: It isn't clear to my why we need the `producer` it is used only to pass to the `RecordCollectorImp` that was a ctor arg previously, and to call `close` on. Why can't we just leave it as it was and call `close` on the `RecordCollector`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Why is this not private?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: nit: it is `null` by default

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I'd probably extract this into a method `getOrCreateProducer()` or similar

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: This is a big code smell IMO. Using reflection to access fields in a class you are testing should be telling you something about the code. I'd prefer to see the `RecordCollector` back to being a ctor argument.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Is this the reason that `threadProducer` is not `private`. I'm not 100% sure, but if it is maybe there is another way of testing this so that it can be `private`, i.e., if you are using the per task model you know that the `clientSupplier` should be called n times. However, with the thread model it should only be called once.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:97]: I did some clean up on the side... Sorry. :(

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Yes. I like to add it explicitly -- personal taste. Can remove it if it bothers you :)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: For testing only. I'll try to make it private using your ideas from: https://github.com/apache/kafka/pull/2773#discussion_r109122317

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: I guess that would work, too. I personally think, `RecordCollector` is a "task internal detail" and thus, the task should create the `RecordCollector` instead of requesting it as input. Might be quite subjective; to me, this is the cleaner design and makes the code in `StreamThread#createStreamTask` simpler.

I am open to change it back though.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: That's a fair argument. Will change it back.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: That would not be sufficient IMHO. Added a test that checks if producers are closed correctly, thus, need access to each producer. Nevertheless, did rework the code (producer member is private plus new non-private method that does return the producer)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: It is a bit subjective. Though as the producer is only used as an argument to the `RecordCollector`, IMO the `RecordCollector` is the right dependency.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: I'm not that bothered, just something we generally don't do and perhaps we should stick with it? either way i'm ok (nit anyway!)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: FWIW - this is probably something that could be tested easily with good usage of a real Mocking framework (rather than hand coded stubs). We should probably start making use of something in streams as it would save a lot of effort and with correct usage lead to  a better overall design.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: Hmmm... Anyway -- need to change it back to avoid reflection in the test.

**Comment by mjsax** [comment]
Updated.

**Comment by mjsax** [comment]
Retest this please.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:401]: Should we extract these at into `public static final fields`? Can then refer to them elsewhere, i.e., in `StreamThread`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:402]: Is this necessary? There is a guard here `if(exactlyOnceEnabled)` there is also the same in `StreamThread` line 422. Could we remove the one from `StreamThread` and just not throw the `IllegalStateException` ? What is it giving us?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:427]: see comment on `StreamTask` do we need to have a guard both here and in `StreamTask#closeProducer`?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java]: `shouldCloseProducerWhenExactlyOneEnabled` ?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: `shouldUseProducerPerThreadWhenEOSDisabled` ?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: looks like this isn't used

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:97]: Ideally we do the cleanup in a second commit (same PR) to make the review easier. At least for me.

**Comment by enothereska** [comment]
Could we have 1 run of the benchmarks using the branch builder  on tests/kafkatest/benchmarks/streams/? Thanks @mjsax

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:97]: I basically agree -- but because this change is trivial, I though is should be ok. I don't want to do an extra PR for the cleanup...

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:401]: I did not do that, to align with ofter config classes. For example, consumer dose not offer "earliest" / "latest" / "none" either. I think consistency is more important at the moment. If we really switch to the builder pattern, this will resolve naturally. WDYT? I am open to change it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:402]: It's a check for bugs -- it should never happen that `closeProducer` get's called. I can also make it an assertion is you like this better.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:427]: The check here is mandatory -- the other check has only the purpose to guard against bugs.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: This tests that Streams does not call `new Client` but used the `ClientSupplier`. We could split it into two tests though -- one for non-eos and one for eos. Not sure if splitting give much value?

**Comment by mjsax** [comment]
Triggered https://jenkins.confluent.io/job/system-test-kafka-branch-builder/823/

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:401]: yeah sure

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: I thought you already had a similar test that tests that when eos is enabled `ClientSupplier` is called multiple times?

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java]: Ah -- forgot that I did the split already -- haha -- too much stuff going on atm... Will do the renaming :)

**Comment by mjsax** [comment]
Wrong test setup. Run run: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/824/

**Review by dguy — APPROVED** [review]
Thanks @mjsax, LGTM

**Comment by enothereska** [comment]
LGTM, I'm happy that perf numbers haven't changed. Ideally we run rest of streams tests under kafkatest/tests/streams however I'm aware that the branch builder is currently clogged.

**Comment by mjsax** [comment]
Triggered: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/831/

**Comment by mjsax** [comment]
Not sure why it did not find the branch... triggered a new build: Triggered: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/834/

**Comment by mjsax** [comment]
Again: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/251/

**Comment by enothereska** [comment]
LGTM thanks!

**Comment by mjsax** [comment]
Call for final review and merging @ijuma

**Comment by ijuma** [comment]
Retest this please

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (ijuma) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:401]: That's a bad pattern in `ConsumerConfig` that we should not be trying to emulate. We also have the following which doesn't use magic values:

```
in(Sensor.RecordingLevel.INFO.toString(), Sensor.RecordingLevel.DEBUG.toString()),
```

Is the builder pattern switch happening in this release?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:718]: `assertEquals` should be used here and any other place where we are comparing numbers. `assertSame` checks for reference equality and it may return false if the number is outside the cacheable range even if they're equal (the various `valueOf` methods used by auto-boxing cache numbers within a certain range).

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:766]: Something to consider for a future PR: it's a bit odd that `MockClientSupplier` always returns the same producer when the contract is that `getProducer` returns a new producer. If we changed it so that it had the specified behaviour we would not need this class.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:1276]: Why don't we just fix `MockProducer` instead of introducing a new class?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:605]: Why don't we just fix MockProducer instead of introducing a new class?

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:402]: Is this MEDIUM or HIGH?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:458]: This cast seems suboptimal. Seems like it would be better to return the `recordCollector` and let the test or a test utility cast, if necessary. Non-test code shouldn't be assuming a particular implementation IMO (yes, I now that the method is for testing only, but it's better to limit such methods to providing access only).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:394]: Would it not be better for this to be called `close`? It's odd to have a `closeProducer` method that calls `recordCollector.close`. It seems like a layering violation.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (ijuma) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:402]: I think we are making things more coupled than they need to be. Why don't we introduce the notion of `StreamsTask.close`. This method should always be called. In `RecordCollectorImpl`, we take a boolean that indicates whether it owns the Producer or not. If it does, it closes the producer in the `close` method. If not, it doesn't. This way `StreamsTask` doesn't even need to care about the `exactlyOnce` setting.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:489]: Similar to my other comment, it seems that we should be thinking about closing tasks. That way, if we have other resources associated with a task in the future, we don't have to change any code.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, I left a few suggestions.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:402]: Probably HIGH

**Comment by enothereska** [comment]
@ijuma raises valid points.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:402]: I think MEDIUM is fine -- my interpretation is: HIGH -> you must most likely change it; MEDIUM -> double check, but default is probably ok; LOW: default is most likely ok -- only consider if you have a good reason

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:401]: I doubt it -- not enough time to do the builder pattern.

**Comment by mjsax** [comment]
About closing producers: the problem is, that we first close all tasks, than we flush all state (and still need the producer for this). Thus, we need to decouple closing tasks and closing producers. If you have any better idea (and introducing method 'closeProducers`) to make the code cleaner I am happy to do so.... \cc @ijuma @enothereska @dguy

**Comment by mjsax** [comment]
We still can change the code, to let `RecordCollector` close the producer instead of the task -- but it does not really simplify it. IMHO, `RecordCollector` should be agnostic and only `StreamTask` should know about the difference of "at-least-once/exactly-once".

**Comment by mjsax** [comment]
We also do have `StreamTask#closeTopology()` for keeping tasks open during rebalance. I guess, we could change the pattern to `closeAllTopologies()` - `closeStateStores` - `closeAllTasks` ?

**Comment by ijuma** [comment]
I merged this to trunk after fixing `assertSame` to `assertEquals` to unblock other work and because other reviewers were happy with the original solution. I still think the layering is not right and abstractions seem to be leaking, but I'll leave it to you guys to decide if it's worth cleaning that up in a subsequent PR.

**Comment by mjsax** [comment]
@ijuma Oh. Wow. Did not expect that. Will do a follow up PR...

**Comment by ijuma** [comment]
To clarify: since these methods are not public and more work is being done for KIP-129, it seemed OK to iterate via subsequent PRs instead of blocking this one.

**Comment by mjsax** [comment]
Follow up PR is #2854

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:97]: +1 to @enothereska

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:113]: Why not remove `this.` on line 114 as well?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:402]: +1

**Pull request closed without merging**
